### DISCUSSION TITLE
Required properties with empty strings

### DISCUSSION THREAD

**Discussion by mdesousa** [vazco/uniforms#1002] [Q&A] [answered]
Hi! I'm creating an AutoField and noticed that a user can delete a value for a required field, and no error is detected.
This is when using json-schema... the value is set to an empty string, which strictly speaking is a value but not the behavior that we want for a form.
Any suggestions on how to handle this easily and globally?
We are using meterial ui... thanks!

**Comment by radekmie** [comment, marked as answer]
Hi @mdesousa. The thing is, that an empty string **is** a valid string. It's up to you to decide what should the schema check for. I'd suggest one of the following, in the order of my personal preference:
1. Use [`pattern`](https://json-schema.org/understanding-json-schema/reference/string.html#regular-expressions) or [`format`](https://json-schema.org/understanding-json-schema/reference/string.html#format) to specify the... Well, pattern or format of the string.
2. Use [`minLength: N`](https://json-schema.org/understanding-json-schema/reference/string.html#length) to specify the minimum input length.
3. Create a custom keyword like `isNotEmpty`.

Please note that all of the above require changing your schema. That's not necessarily bad, as one, this can be (semi-)automated, and clearly defines your intention.

Also, remember that there's no trimming by default, and you may want to use a custom keyword for that as well. Luckily, `ajv-keywords` got you covered with its [`transform` keyword](https://github.com/ajv-validator/ajv-keywords#transform).

**Reply by mdesousa** [reply]
Thanks @radekmie these are good suggestions! Yes I understand that the difference is having `undefined` or `null` vs `""`.
However it is weird that a field for a newly initialized object has an undefined value, and when the user touches and deletes the field you get an empty string. Was wondering if there was an easy way to automatically represent that scenario as `undefined`.

**Reply by radekmie** [reply]
Yes, that's somehow problematic, I know. To be honest no, I don't see an easy solution here. One thing that I often saw in a couple of projects was a "clean" phase before validation, that'd remove empty strings and arrays, etc.

**Reply by Krypt0r7** [reply]
Hi @radekmie and @mdesousa!

This is exactly how we solved the issue. We had problems with some Objects and Lists that when first had a value that then where removed/deleted still was an empty List/Object. This was problematic due to server side validations and the need to keep some properties as null. 

We used Lodash to work through the model to set all empty properties to null before submit to server.

And receiving a value that is null (lists, objects) makes the form not editable.

To me the expected behaviour should be to remove the property if it is empty (mostly Arrays) or at least set it to null.

**Reply by radekmie** [reply]
>And receiving a value that is null (lists, objects) makes the form not editable.
> [...] or at least set it to null.

But does your schema actually allow `null`s in these values? The problem here is that we cannot allow `null`s everywhere, because often it's against the schema. And simply removing the values _is_ a valid approach, yet we haven't gone that way. We may consider that for the future (at least v4 as it's a breaking change).

**Reply by mdesousa** [reply]
hi @radekmie what is the best way to do a "clean" of a field before validation?

**Reply by radekmie** [reply]
@mdesousa you can use `modelTransform` for that, or do it directly in the validator.
